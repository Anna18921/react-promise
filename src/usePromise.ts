import { useEffect, useState } from 'react'

export default function usePromise<T>(
  promiseOrFn: (() => Promise<T>) | Promise<T>
): {
  loading: boolean
  error: Error | null
  value: T | undefined
} {
  const [state, setState] = useState<{
    loading: boolean
    error: Error | null
    value: T | undefined
  }>({
    loading: !!promiseOrFn,
    error: null,
    value: undefined
  })

  useEffect(() => {
    let unmounted = false

    if (!promiseOrFn) {
      setState({
        loading: false,
        error: null,
        value: undefined
      })
    } else {
      if (state.loading === false) {
        setState({
          loading: true,
          error: null,
          value: undefined
        })
      }
      let promise: Promise<T>
      if (typeof promiseOrFn === 'function') {
        promise = promiseOrFn()
      } else {
        promise = promiseOrFn
      }

      promise
        .then((value) => {
          if (!unmounted) {
            setState({
              loading: false,
              error: null,
              value
            })
          }
        })
        .catch((error) => {
          if (!unmounted) {
            setState({
              loading: false,
              error,
              value: undefined
            })
          }
        })
    }

    return () => {
      unmounted = true
    }
  }, [promiseOrFn])

  return state
}
