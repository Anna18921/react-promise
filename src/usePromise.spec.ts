import usePromise from './usePromise'
import { renderHook, act } from 'react-hooks-testing-library'

describe('usePromise', () => {
  it('works for resolved promise', (done) => {
    let prom = new Promise((resolve) => {
      setTimeout(function() {
        act(() => {
          resolve('a value')
        })
      }, 49)
    })

    renderHook(() => {
      act(() => {
        expect(usePromise(prom)).toMatchSnapshot()
      })

      setTimeout(() => {
        done()
      }, 50)
    })
  })

  it('works for rejected promise', (done) => {
    let prom = new Promise((_, reject) => {
      setTimeout(function() {
        act(() => {
          reject(new Error('an error'))
        })
      }, 49)
    })
    renderHook(() => {
      act(() => {
        expect(usePromise(prom)).toMatchSnapshot()
      })

      setTimeout(() => {
        done()
      }, 50)
    })
  })

  it('works when providing a promise thunk', (done) => {
    let promThunk = () =>
      new Promise((resolve) => {
        setTimeout(function() {
          act(() => {
            resolve('a value from promise returning thunk')
          })
        }, 49)
      })

    renderHook(() => {
      act(() => {
        expect(usePromise(promThunk)).toMatchSnapshot()
      })

      setTimeout(() => {
        done()
      }, 50)
    })
  })
})
