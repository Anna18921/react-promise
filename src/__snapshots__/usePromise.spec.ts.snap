// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`usePromise works for rejected promise 1`] = `
Object {
  "error": null,
  "loading": true,
  "value": undefined,
}
`;

exports[`usePromise works for rejected promise 2`] = `
Object {
  "error": [Error: an error],
  "loading": false,
  "value": undefined,
}
`;

exports[`usePromise works for resolved promise 1`] = `
Object {
  "error": null,
  "loading": true,
  "value": undefined,
}
`;

exports[`usePromise works for resolved promise 2`] = `
Object {
  "error": null,
  "loading": false,
  "value": "a value",
}
`;

exports[`usePromise works when providing a promise thunk 1`] = `
Object {
  "error": null,
  "loading": true,
  "value": undefined,
}
`;

exports[`usePromise works when providing a promise thunk 2`] = `
Object {
  "error": null,
  "loading": false,
  "value": "a value from promise returning thunk",
}
`;
